dbcon/mysql/ha_mcs_pushdown.cpp

```diff
@@ -0,0 +1,620 @@
+/*
+   Copyright (c) 2019 MariaDB
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+
+// ha_calpont.cpp includes this file.
+
+void mutate_optimizer_flags(THD *thd_)
+{
+   // MCOL-2178 Disable all optimizer flags as it was in the fork.
+    // CS restores it later in SH::scan_end() and in case of an error
+    // in SH::scan_init()
+    set_original_optimizer_flags(thd_->variables.optimizer_switch, thd_);
+    thd_->variables.optimizer_switch = OPTIMIZER_SWITCH_IN_TO_EXISTS |
+        OPTIMIZER_SWITCH_EXISTS_TO_IN |
+        OPTIMIZER_SWITCH_COND_PUSHDOWN_FOR_DERIVED;
```

Comment:
We need to rethink this. We actually only had one or two flags that didn't work but ended up switching most of them off in a massive hammer fashion. We should probably fix this.

Reply:
I would better disable all the flags and later enable them one by one. But this work is big enough for a separate issue.

Reply:
Completely agree